internal/libyaml/yamldatatest_loader.go

```diff
@@ -0,0 +1,503 @@
+
+		case MAPPING_END_EVENT:
+			if len(stack) > 0 {
+				popped := stack[len(stack)-1]
+				stack = stack[:len(stack)-1]
+
+				// Add completed map to parent or set as root
+				if len(stack) == 0 {
+					root = popped.container
+				} else {
+					parent := &stack[len(stack)-1]
+					if m, ok := parent.container.(map[string]interface{}); ok {
+						m[parent.key] = popped.container
+						parent.key = "" // Reset key after use
+					} else if s, ok := parent.container.([]interface{}); ok {
+						parent.container = append(s, popped.container)
+					}
+				}
+			}
+
+		case SEQUENCE_START_EVENT:
+			newSlice := make([]interface{}, 0)
+			stack = append(stack, stackEntry{container: newSlice})
+
+		case SEQUENCE_END_EVENT:
+			if len(stack) > 0 {
+				popped := stack[len(stack)-1]
+				stack = stack[:len(stack)-1]
+
+				// Add completed slice to parent or set as root
+				if len(stack) == 0 {
+					root = popped.container
+				} else {
+					parent := &stack[len(stack)-1]
+					if m, ok := parent.container.(map[string]interface{}); ok {
+						m[parent.key] = popped.container
+						parent.key = "" // Reset key after use
+					} else if s, ok := parent.container.([]interface{}); ok {
+						parent.container = append(s, popped.container)
+					}
+				}
+			}
+
+		case SCALAR_EVENT:
+			value := string(event.Value)
+			// Only coerce plain (unquoted) scalars
+			isQuoted := ScalarStyle(event.Style) != PLAIN_SCALAR_STYLE
+
+			if len(stack) == 0 {
+				// Scalar at root level
+				if isQuoted {
+					root = value
+				} else {
+					root = coerceScalar(value)
+				}
+			} else {
+				parent := &stack[len(stack)-1]
+				if m, ok := parent.container.(map[string]interface{}); ok {
+					if parent.key == "" {
+						// This scalar is a key - keep as string, don't coerce
+						parent.key = value
+					} else {
+						// This scalar is a value
+						if isQuoted {
+							m[parent.key] = value
+						} else {
+							m[parent.key] = coerceScalar(value)
+						}
+						parent.key = ""
+					}
+				} else if s, ok := parent.container.([]interface{}); ok {
+					// Add to sequence
+					if isQuoted {
+						parent.container = append(s, value)
+					} else {
+						parent.container = append(s, coerceScalar(value))
+					}
+				}
+			}
+
+		case DOCUMENT_END_EVENT:
+			// Document end marker, continue processing
+
+		case ALIAS_EVENT, TAIL_COMMENT_EVENT:
+			// For now, skip aliases and comments (not used in test data)
+		}
+	}
+
+	return root, nil
+}
+
+// UnmarshalStruct populates a struct from a map using reflection and yaml tags.
+// This function is exported so it can be used by other packages for data-driven testing.
+//
+// Example:
+//
+//	type TestCase struct {
+//	    Name string `yaml:"name"`
+//	    Data string `yaml:"data"`
+//	}
+//	var tc TestCase
+//	err := libyaml.UnmarshalStruct(&tc, map[string]interface{}{
+//	    "name": "test1",
+//	    "data": "hello",
+//	})
+func UnmarshalStruct(target interface{}, data map[string]interface{}) error {
```

Comment:
I get why you're doing this and it does make sense. Are we worried about any possibility of this becoming misaligned with the root level method?